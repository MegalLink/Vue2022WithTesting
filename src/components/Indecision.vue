<template>
  <img v-if="img" :src="img" alt="bg" />
  <div class="bg-dark"></div>
  <div class="container-indecision">
    <input v-model="question" type="text" placeholder="Que quieres saber?" />
    <p>Porfavor introduzca una pregunta</p>
    <div v-if="isValidQuestion">
      <h2>{{ question }}</h2>
      <h1>{{ answer }}</h1>
    </div>
  </div>
</template>
<script>
export default {
  data() {
    return {
      question: null,
      answer: null,
      img: null,
      isValidQuestion: false,
    };
  },
  methods: {
    async getAnswer() {
      try {
        this.answer = "Pensando....";
        const { answer, image } = await fetch("https://yesno.wtf/api").then(
          (r) => r.json()
        );
        switch (answer) {
          case "yes":
            this.answer = "Si :D";
            break;
          case "no":
            this.answer = "No :(";
            break;
          default:
            this.answer = "No s√© ._.";
        }

        this.img = image;
      } catch (error) {
        this.answer = "No hay respues del API";
        this.img = null;
      }
    },
  },
  watch: {
    question(newVal, _) {
      this.isValidQuestion = false;
      console.log(newVal);
      if (newVal.includes("?")) {
        this.getAnswer();
        this.isValidQuestion = true;
      }
    },
  },
};
</script>
<style scoped>
img,
.bg-dark {
  height: 100vh;
  left: 0px;
  max-height: 100%;
  max-width: 100%;
  position: fixed;
  top: 0px;
  width: 100vw;
}

.bg-dark {
  background-color: rgba(0, 0, 0, 0.4);
}

.container-indecision {
  position: relative;
  z-index: 99;
}

input {
  width: 250px;
  padding: 10px 15px;
  border-radius: 5px;
  border: none;
}
input:focus {
  outline: none;
}

p {
  color: white;
  font-size: 20px;
  margin-top: 0px;
}

h1,
h2 {
  color: white;
}

h2 {
  margin-top: 150px;
}
</style>
