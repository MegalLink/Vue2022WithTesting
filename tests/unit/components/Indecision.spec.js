import Indecision from "@/components/Indecision";
import { shallowMount } from "@vue/test-utils";

describe("Indecision Component", () => {
  let wrapper;
  let consoleLogSpy;
  global.fetch = jest.fn();
  beforeEach(() => {
    wrapper = shallowMount(Indecision);
    consoleLogSpy = jest.spyOn(console, "log");
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  function setMockFetch(response = null, isOk = false) {
    return () => {
      if (isOk) {
        return Promise.resolve({
          json: () => Promise.resolve(response),
        });
      } else {
        return Promise.reject({
          json: () =>
            Promise.reject({
              statusCode: 500,
            }),
        });
      }
    };
  }

  it("should match snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should not trigger anything when answer dont have '?'", async () => {
    const input = wrapper.find("input");
    //THIS VM has mock of functions, getter and setter for data
    const getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer");
    await input.setValue("Hola mundo");
    expect(consoleLogSpy).toHaveBeenCalled();
    expect(getAnswerSpy).not.toHaveBeenCalled();
  });

  it("should trigger getAnswer when answer have '?'", async () => {
    const input = wrapper.find("input");
    //THIS VM has mock of functions, getter and setter for data
    const getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer");
    await input.setValue("Hola mundo?");
    getAnswerSpy.mockResolvedValue("yes");
    //fetch is not defined in node so we have to mock it...

    expect(consoleLogSpy).toHaveBeenCalled();
    expect(getAnswerSpy).toHaveBeenCalled();
  });

  it("should return reponse when getAnswer method is called and answer is yes", async () => {
    const urlImage = "urlimage";
    const apiResponse = { answer: "yes", forced: false, image: urlImage };
    fetch.mockImplementationOnce(setMockFetch(apiResponse, true));
    await wrapper.vm.getAnswer();
    const img = wrapper.find("img");
    expect(wrapper.vm.img).toBe(urlImage);
    expect(img.exists()).toBeTruthy();
    expect(wrapper.vm.answer).toBe("Si :D");
  });

  it("should return reponse when getAnswer method is called and answer is no", async () => {
    const urlImage = "urlimage";
    const apiResponse = { answer: "no", forced: false, image: urlImage };
    fetch.mockImplementationOnce(setMockFetch(apiResponse, true));
    await wrapper.vm.getAnswer();
    const img = wrapper.find("img");
    expect(wrapper.vm.img).toBe(urlImage);
    expect(img.exists()).toBeTruthy();
    expect(wrapper.vm.answer).toBe("No :(");
  });

  it("should return reponse when getAnswer method is called and answer is other answer", async () => {
    const urlImage = "urlimage";
    const apiResponse = {
      answer: "i dont know",
      forced: false,
      image: urlImage,
    };
    fetch.mockImplementationOnce(setMockFetch(apiResponse, true));
    await wrapper.vm.getAnswer();
    const img = wrapper.find("img");
    expect(wrapper.vm.img).toBe(urlImage);
    expect(img.exists()).toBeTruthy();
    expect(wrapper.vm.answer).toBe("No sÃ© ._.");
  });

  it("should return error when getAnswer method is called", async () => {
    fetch.mockImplementationOnce(setMockFetch());
    await wrapper.vm.getAnswer();
    const img = wrapper.find("img");
    expect(wrapper.vm.img).toBe(null);
    expect(img.exists()).toBeFalsy();
    expect(wrapper.vm.answer).toBe("No hay respues del API");
  });
});
