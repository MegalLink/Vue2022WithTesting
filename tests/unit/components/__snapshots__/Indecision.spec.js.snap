// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision Component should match snapshot 1`] = `
<!--v-if-->
<div class="bg-dark"></div>
<div class="container-indecision"><input type="text" placeholder="Que quieres saber?">
  <p>Porfavor introduzca una pregunta</p>
  <!--v-if-->
</div>
`;
